service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageDocumentService.java

```diff
@@ -0,0 +1,232 @@
+        C2DocumentBundle c2DocumentBundle = caseData.getC2DocumentBundleByUUID(selectedC2);
+
+        if (c2DocumentBundle.getSupportingEvidenceBundle() != null) {
+            return c2DocumentBundle.getSupportingEvidenceBundle();
+        }
+
+        return getSupportingEvidenceBundle(null);
+    }
+
+    public List<Element<SupportingEvidenceBundle>> getSupportingEvidenceBundle(
+        List<Element<SupportingEvidenceBundle>> supportingEvidenceBundleList) {
+        if (supportingEvidenceBundleList == null || supportingEvidenceBundleList.isEmpty()) {
+            return List.of(element(SupportingEvidenceBundle.builder().build()));
+        }
+
+        return supportingEvidenceBundleList;
+    }
+
+    public List<Element<HearingFurtherEvidenceBundle>> buildHearingFurtherEvidenceCollection(
+        CaseData caseData, List<Element<SupportingEvidenceBundle>> supportingEvidenceBundle) {
+
+        List<Element<HearingFurtherEvidenceBundle>> hearingFurtherEvidenceBundle;
+
+        hearingFurtherEvidenceBundle = caseData.getHearingFurtherEvidenceDocuments();
+        UUID selectedHearingCode = getDynamicListValueCode(caseData.getManageDocumentsHearingList(), mapper);
+        HearingBooking hearingBooking = getHearingBookingByUUID(caseData.getHearingDetails(), selectedHearingCode);
+
+        if (caseData.documentBundleContainsHearingId(selectedHearingCode)) {
+            List<Element<HearingFurtherEvidenceBundle>> updateEvidenceBundles = new ArrayList<>();
+            for (Element<HearingFurtherEvidenceBundle> element : hearingFurtherEvidenceBundle) {
+                if (element.getId().equals(selectedHearingCode)) {
+                    element.getValue().setSupportingEvidenceBundle(supportingEvidenceBundle);
+                }
+                updateEvidenceBundles.add(element);
+            }
+            return updateEvidenceBundles;
+        } else {
+            hearingFurtherEvidenceBundle.add(buildHearingSupportingEvidenceBundle(
+                selectedHearingCode,
+                hearingBooking,
+                supportingEvidenceBundle
+            ));
+            return hearingFurtherEvidenceBundle;
+        }
+    }
+
+    public List<Element<SupportingEvidenceBundle>> setDateTimeUploadedOnSupportingEvidence(
+        List<Element<SupportingEvidenceBundle>> supportingEvidenceBundle,
+        List<Element<SupportingEvidenceBundle>> supportingEvidenceBundleBefore) {
+
+        if (!Objects.equals(supportingEvidenceBundle, supportingEvidenceBundleBefore)) {
+            List<Element<SupportingEvidenceBundle>> altered = new ArrayList<>(supportingEvidenceBundle);
+
+            // Could be null in the case of C2 supporting documents
+            if (supportingEvidenceBundleBefore != null) {
+                altered.removeAll(supportingEvidenceBundleBefore);
+            }
+
+            altered.forEach(bundle -> findElement(bundle.getId(), supportingEvidenceBundleBefore).ifPresent(
+                previousVersion -> {
```

Comment:
That could be extracted to separate method - ifPresent(this::updateUploadTimeIfDocument) changed